pkg/mcp/server/server.go

```diff
@@ -87,6 +87,11 @@ type Server struct {
 // this case and apply appropriate policy.
 type AuthChecker interface {
 	Check(authInfo credentials.AuthInfo) error
+	// ShouldRecordFailure is to rate limit auth check failures
+	// to avoid flooding logs with duplicate information.
+	// it returns with bool and int value that indicate failure counts
+	// since last time.
+	ShouldRecordFailure() (bool, int)
```

Comment:
1) We should use https://godoc.org/golang.org/x/time/rate instead of re-implementing our own rate limiting code
2) We don't need to add `ShouldRecordFailure()` method to the `AuthChecker` interface. The frequency of logging auth check failures should be common to any auth checker and can be implemented in the `newConnection` server method.

Reply:
> We should use https://godoc.org/golang.org/x/time/rate instead of re-implementing our own rate limiting code

The limiter limits number of events *per second*, won't it be too frequent for us here?

Reply:
Limit is a float allowing for a single event to be spread across multiple seconds, e.g.

```golang
$ cat limit.go 
package main

import (
        "fmt"
        "time"

        "golang.org/x/time/rate"
)

func main() {
        limit := rate.Every(2 * time.Second)
        fmt.Println("Limit is", limit)
        limiter := rate.NewLimiter(limit, 1)
        for i := 0; ; i++ {
                if limiter.Allow() {
                        fmt.Printf("[%v] Allow()=true for iteration %v \n", time.Now().UTC(), i)
                }
                time.Sleep(time.Millisecond)
        }
}

$ go run limit.go 
Limit is 0.5
[2018-09-07 20:59:54.739810365 +0000 UTC] Allow()=true for iteration 0 
[2018-09-07 20:59:56.740787876 +0000 UTC] Allow()=true for iteration 1739 
[2018-09-07 20:59:58.740851503 +0000 UTC] Allow()=true for iteration 3449 
[2018-09-07 21:00:00.741663849 +0000 UTC] Allow()=true for iteration 524
```